File: libs/power-log-parser/src/lib/parsers/battlegrounds-reward-gained-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Node, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class BattlegroundsRewardGainedParser implements ActionParser {
	readonly ParserName = 'BattlegroundsRewardGainedParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			this.StateFacade.IsBattlegrounds() &&
			node.Type === TagChange &&
			((node.Object as TagChange).Name === (GameTag.BACON_HERO_QUEST_REWARD_COMPLETED as number) ||
				(node.Object as TagChange).Name === (GameTag.BACON_HERO_HEROPOWER_QUEST_REWARD_COMPLETED as number))
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const hero = this.GameState.CurrentEntities.get(tagChange.Entity);
		if (hero?.CardId == null) {
			return null;
		}

		const isHeroPowerReward =
			tagChange.Name === (GameTag.BACON_HERO_HEROPOWER_QUEST_REWARD_COMPLETED as number);
		const questRewardDbfId = isHeroPowerReward
			? hero.GetTag(GameTag.BACON_HERO_HEROPOWER_QUEST_REWARD_DATABASE_ID)
			: hero.GetTag(GameTag.BACON_HERO_QUEST_REWARD_DATABASE_ID);

		const controllerId = hero.GetEffectiveController();
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'BATTLEGROUNDS_REWARD_GAINED',
				GameEventHelper.CreateProvider(
					'BATTLEGROUNDS_REWARD_GAINED',
					hero.CardId,
					controllerId,
					hero.Entity,
					this.StateFacade,
					{
						PlayerId: hero.GetTag(GameTag.PLAYER_ID),
						QuestRewardDbfId: questRewardDbfId,
						IsHeroPowerReward: isHeroPowerReward,
					},
				),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BattlegroundsRewardGainedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Linq;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using static HearthstoneReplays.Events.CardIds;

namespace HearthstoneReplays.Events.Parsers
{
    public class BattlegroundsRewardGainedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public BattlegroundsRewardGainedParser(ParserState ParserState, StateFacade stateFacade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = stateFacade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && StateFacade.IsBattlegrounds()
                && node.Type == typeof(TagChange)
                && ((node.Object as TagChange).Name == (int)GameTag.BACON_HERO_QUEST_REWARD_COMPLETED
                    // For Denathrius
                    || (node.Object as TagChange).Name == (int)GameTag.BACON_HERO_HEROPOWER_QUEST_REWARD_COMPLETED);
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var hero = GameState.CurrentEntities.GetValueOrDefault(tagChange.Entity);
            if (hero?.CardId == null)
            {
                return null;
            }

            // TODO: find the reward card id
            var isHeroPowerReward = tagChange.Name == (int)GameTag.BACON_HERO_HEROPOWER_QUEST_REWARD_COMPLETED;
            var questRewardDbfId = isHeroPowerReward ? hero.GetTag(GameTag.BACON_HERO_HEROPOWER_QUEST_REWARD_DATABASE_ID) : hero.GetTag(GameTag.BACON_HERO_QUEST_REWARD_DATABASE_ID);

            var controllerId = hero.GetEffectiveController();
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                "BATTLEGROUNDS_REWARD_GAINED",
                GameEvent.CreateProvider(
                "BATTLEGROUNDS_REWARD_GAINED",
                    hero.CardId,
                    controllerId,
                    hero.Entity,
                    StateFacade,
                    //null,
                    new {
                        PlayerId = hero.GetTag(GameTag.PLAYER_ID),
                        QuestRewardDbfId = questRewardDbfId,
                        IsHeroPowerReward = isHeroPowerReward,
                    }),
                true,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
